File: packages/provide-uterm-go/hub/resthelpers.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

import (
	"encoding/json"
	"fmt"
	"regexp"
)

// maxExpectRegexLen is the maximum allowed prompt-guard regex length. Port of
// rest_helpers.MAX_EXPECT_REGEX_LEN.
const maxExpectRegexLen = 200

// PromptRegexError is returned when a prompt guard regex is invalid, unsafe, or
// too long. Port of rest_helpers.PromptRegexError. Kind is one of "too_long",
// "unsafe", "invalid".
type PromptRegexError struct {
	Message string
	Kind    string
}

func (e *PromptRegexError) Error() string { return e.Message }

// extractPromptID pulls prompt_detected.prompt_id from a snapshot, or "" when
// absent/malformed. Port of rest_helpers.extract_prompt_id (Go "" == Python None).
func extractPromptID(snapshot map[string]any) string {
	if len(snapshot) == 0 {
		return ""
	}
	prompt, ok := snapshot["prompt_detected"].(map[string]any)
	if !ok {
		return ""
	}
	value, ok := prompt["prompt_id"].(string)
	if ok && value != "" {
		return value
	}
	return ""
}

// compileExpectRegex compiles a prompt guard regex (with the IGNORECASE +
// MULTILINE flags applied via the (?im) prefix) or returns a *PromptRegexError.
// Port of rest_helpers.compile_expect_regex. An empty regex returns (nil, nil).
func compileExpectRegex(expectRegex string) (*regexp.Regexp, error) {
	if expectRegex == "" {
		return nil, nil
	}
	if len(expectRegex) > maxExpectRegexLen {
		return nil, &PromptRegexError{Message: "expect_regex too long", Kind: "too_long"}
	}
	if err := validatePatternSafety(expectRegex); err != nil {
		return nil, &PromptRegexError{Message: fmt.Sprintf("unsafe expect_regex: %s", err.Error()), Kind: "unsafe"}
	}
	re, err := regexp.Compile("(?im)" + expectRegex)
	if err != nil {
		return nil, &PromptRegexError{Message: fmt.Sprintf("invalid expect_regex: %s", err.Error()), Kind: "invalid"}
	}
	return re, nil
}

// snapshotMatches reports whether snapshot satisfies the prompt-id and/or regex
// guard. Port of rest_helpers.snapshot_matches.
func snapshotMatches(snapshot map[string]any, expectPromptID string, expectRegex *regexp.Regexp) bool {
	if snapshot == nil {
		return false
	}
	if expectPromptID != "" && extractPromptID(snapshot) != expectPromptID {
		return false
	}
	return expectRegex == nil || expectRegex.MatchString(toStr(snapshot["screen"]))
}

// toStr coerces a value to the string form Python's str() would produce for the
// inputs the ported code encounters: a string is returned verbatim, a missing
// (nil) value yields "", and anything else is fmt.Sprint'd.
func toStr(v any) string {
	if v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprint(v)
}

// coerceFloat coerces an int/float value to float64, returning def otherwise.
// json.Number is handled because controlchannel decodes wire numbers into it
// (preserving the int/float text form for signature-faithful round-trips).
func coerceFloat(v any, def float64) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	case int64:
		return float64(n)
	case json.Number:
		f, err := n.Float64()
		if err != nil {
			return def
		}
		return f
	default:
		return def
	}
}

File: packages/provide-uterm-ts/src/hub/rest-helpers.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Prompt guards for the hijack REST contract.
 *
 * Port of the Python module `provide.uterm.server.bridge.rest_helpers` and
 * the Go package `hub` (`resthelpers.go`).
 *
 * A REST caller attaches these to a keystroke send: "type this, then wait
 * until the screen looks like *that*". Both failure directions cost the
 * caller — a guard that matches too eagerly returns before the command has
 * run, and one that never matches hangs until the timeout.
 */

import { validatePatternSafety } from "./pattern-safety.ts";

/**
 * Longest guard pattern a caller may supply.
 *
 * The bound is about work, not storage: this is searched against a whole
 * screen on every poll.
 */
export const MAX_EXPECT_REGEX_LEN = 200;

/** Why a guard pattern was refused. */
export type PromptRegexErrorKind = "too_long" | "unsafe" | "invalid";

/** Raised when a prompt guard pattern is refused. */
export class PromptRegexError extends Error {
  /** Which rule refused it; callers surface this. */
  readonly kind: PromptRegexErrorKind;
  /** The limit in force, when the refusal was about length. */
  readonly maxLength: number | undefined;

  constructor(message: string, kind: PromptRegexErrorKind, maxLength?: number) {
    super(message);
    this.name = "PromptRegexError";
    this.kind = kind;
    this.maxLength = maxLength;
  }
}

/** Options for {@link compileExpectRegex}. */
export interface CompileExpectRegexOptions {
  /** Overrides {@link MAX_EXPECT_REGEX_LEN}. */
  maxLength?: number;
}

/** Options for {@link snapshotMatches}. */
export interface SnapshotGuards {
  /** Require the detected prompt to have this id. */
  expectPromptId?: string | undefined;
  /** Require the screen to match this pattern. */
  expectRegex?: RegExp | undefined;
}

/**
 * The detected prompt id in a snapshot, if there is one.
 *
 * Every shape this declines is one a worker can actually send, so it returns
 * nothing rather than throwing on any of them. An empty id is declined too:
 * it would compare equal to an empty guard and match everything.
 */
export function extractPromptId(snapshot?: Record<string, unknown>): string | undefined {
  if (snapshot === undefined) {
    return undefined;
  }
  const detected = snapshot.prompt_detected;
  // The array check mirrors the reference's isinstance(dict); JSON cannot
  // produce an array carrying a prompt_id, so it is faithful rather than
  // load-bearing.
  if (typeof detected !== "object" || detected === null || Array.isArray(detected)) {
    return undefined;
  }
  const value = (detected as Record<string, unknown>).prompt_id;
  return typeof value === "string" && value !== "" ? value : undefined;
}

/**
 * Compile a prompt guard pattern.
 *
 * Refuses in three ways, each with its own kind. The length check runs
 * *first*: an over-long pattern must never reach the safety validator, or a
 * caller could spend the validator's time on a megabyte of input.
 *
 * Compiled case-insensitive and multiline, matching the poll loop's flags —
 * prompt text varies in case, and a guard describes one line of a screen
 * rather than the whole buffer. The `g` flag is deliberately *absent*: a
 * global pattern carries `lastIndex` between calls and would report a miss on
 * the next poll of an unchanged screen.
 *
 * @throws {PromptRegexError} When the pattern is too long, unsafe, or will
 *   not compile.
 */
export function compileExpectRegex(pattern?: string, options: CompileExpectRegexOptions = {}): RegExp | undefined {
  if (pattern === undefined || pattern === "") {
    return undefined;
  }
  const maxLength = options.maxLength ?? MAX_EXPECT_REGEX_LEN;
  if (pattern.length > maxLength) {
    throw new PromptRegexError("expect_regex too long", "too_long", maxLength);
  }
  try {
    validatePatternSafety(pattern);
  } catch (error) {
    throw new PromptRegexError(`unsafe expect_regex: ${(error as Error).message}`, "unsafe", maxLength);
  }
  try {
    return new RegExp(pattern, "im");
  } catch (error) {
    throw new PromptRegexError(`invalid expect_regex: ${(error as Error).message}`, "invalid", maxLength);
  }
}

/**
 * Whether `snapshot` satisfies the given guards.
 *
 * An absent snapshot never matches, even with no guards at all: "nothing to
 * check" is not "everything passed" — the caller is waiting for the worker to
 * say something, not for it to stay silent.
 *
 * An empty prompt id is no constraint rather than a constraint on emptiness.
 * When both guards are given, both must hold.
 */
export function snapshotMatches(snapshot: Record<string, unknown> | undefined, guards: SnapshotGuards): boolean {
  if (snapshot === undefined) {
    return false;
  }
  const { expectPromptId, expectRegex } = guards;
  if (expectPromptId !== undefined && expectPromptId !== "" && extractPromptId(snapshot) !== expectPromptId) {
    return false;
  }
  if (expectRegex === undefined) {
    return true;
  }
  return expectRegex.test(String(snapshot.screen ?? ""));
}
